appengine/src/main/java/com/google/googleinterns/gscribe/services/impl/TokenServiceImpl.java

```diff
@@ -16,30 +16,151 @@
+            InputStream in = ExamResource.class.getResourceAsStream(CREDENTIALS_FILE_PATH);
+            if (in == null) {
+                throw new FileNotFoundException("Resource not found: " + CREDENTIALS_FILE_PATH);
+            }
+            GoogleClientSecrets clientSecrets = GoogleClientSecrets.load(JSON_FACTORY, new InputStreamReader(in));
+            GoogleAuthorizationCodeFlow flow = new GoogleAuthorizationCodeFlow.Builder(HTTP_TRANSPORT, JSON_FACTORY, clientSecrets, SCOPES).setAccessType("offline").build();
+            GoogleAuthorizationCodeTokenRequest tokenRequest = flow.newTokenRequest(authCode);
+            tokenRequest.setRedirectUri(clientSecrets.getWeb().getRedirectUris().get(0));
+            GoogleTokenResponse GoogleTokenResponse = tokenRequest.execute();
+            tokenResponse = new TokenResponse(GoogleTokenResponse.getAccessToken(), GoogleTokenResponse.getRefreshToken(), GoogleTokenResponse.getIdToken());
+        } catch (TokenResponseException e) {
+            throw new InvalidRequestException("Authorization failed");
+        }
+        return tokenResponse;
     }
 
+    /**
+     * Called when access token expires so new access token is required
+     * The user tokens are retrieved from database and refreshed if the access token expires
+     * After getting a GoogleTokenResponse for validation we check that this refreshToken is valid for the main user
+     * For validation verify the IDToken from the token received from refreshToken to get a userID
+     * Compare both userIDs of main request and refreshToken
+     *
+     * @param user ( contains refreshToken )
+     * @throws IOException,GeneralSecurityException ( thrown by NetHttpTransport, GoogleClientSecrets, GoogleTokenResponse or by invalid credentials file  )
+     * @throws InvalidDatabaseDataException         ( if the userID retrieved from refreshing the access token is different than the actual userID,
+     *                                              if the IDToken retrieved from refreshing the token is not verified by googleVerifier
+     *                                              if the refresh token is not correct and TokenRequestException is received )
+     */
     @Override
-    public UserToken generateToken(String IDToken, String authCode) {
-        return null;
+    public void refreshToken(User user) throws IOException, InvalidDatabaseDataException, GeneralSecurityException {
+        try {
+            final JsonFactory JSON_FACTORY = JacksonFactory.getDefaultInstance();
+            final String CREDENTIALS_FILE_PATH = "/credentials.json";
+            InputStream in = TokenServiceImpl.class.getResourceAsStream(CREDENTIALS_FILE_PATH);
+            if (in == null) {
+                throw new FileNotFoundException("Resource not found: " + CREDENTIALS_FILE_PATH);
+            }
+            GoogleClientSecrets clientSecrets = GoogleClientSecrets.load(JSON_FACTORY, new InputStreamReader(in));
```

Comment:
Reading file in each and every request is not a good idea, inject GoogleClientSecrets.

Reply:
Created a new Module for GoogleClientSecrets and NetHttpTransport.